engine/crates/gateway-core/src/lib.rs

```diff
@@ -1,151 +1,127 @@
 use tracing::{info_span, Instrument};
 
 mod admin;
 mod auth;
 mod cache;
 mod executor;
-mod streaming_format;
+mod response;
+pub mod serving;
+mod streaming;
 
-pub use auth::{authorize_request, AuthError};
+pub use auth::{authorize_request, AdminAuthError, AuthError, Authorizer};
 pub use cache::{CacheConfig, CacheControl};
 pub use executor::Executor;
-pub use streaming_format::StreamingFormat;
+pub use response::Response;
+pub use streaming::{encode_stream_response, format::StreamingFormat};
 
-#[async_trait::async_trait]
-pub trait RequestContext: Send + Sync + Sized {
-    fn ray_id(&self) -> &str;
-    // worker requires a 'static future, so there isn't any choice.
-    async fn wait_until(&self, fut: BoxFuture<'static, ()>);
-
-    // Legacy for auth, because it also tries to retrieve it from the query params...
-    fn authorization_header(&self) -> Option<String>;
-    fn header(&self, name: &str) -> Option<String>;
-    fn headers(&self) -> HashMap<String, String>;
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("Bad request: {0}")]
+    BadRequest(String),
+    #[error(transparent)]
+    Cache(runtime::cache::Error),
+    #[error("Serialization Error: {0}")]
+    Serialization(String),
 }
 
-#[async_trait::async_trait]
-pub trait Gateway: Send {
-    // A bit tedious but this allows the caller to have an easier time with its error, response and
-    // context types.
-    type Error: From<Error> + std::error::Error + Send + 'static;
-    type Context: RequestContext;
-    type Response: Response<Context = Self::Context, Error = Self::Error>;
-    type Cache: Cache<Value = engine::Response> + 'static;
-    type Executor: Executor<Context = Self::Context, Error = Self::Error, Response = Self::Response> + 'static;
-
-    fn context(&self) -> &Arc<Self::Context>;
-    fn cache(&self) -> &Arc<Self::Cache>;
-    fn executor(&self) -> &Arc<Self::Executor>;
-    fn cache_config(&self) -> Cow<'_, CacheConfig<'_>>;
-
-    async fn authorize_admin_request(&self, _request: &async_graphql::Request) -> Result<(), AdminAuthError>;
+// A bit tedious but this allows the caller to have an easier time with its error, response and
+// context types.
+pub struct Gateway<Executor, Cache, Context, Response, Error> {
+    executor: Arc<Executor>,
+    cache: Arc<Cache>,
+    cache_config: CacheConfig,
+    authorizer: Box<dyn Authorizer<Context = Context>>,
+    _phantom: PhantomData<fn(Context) -> (Response, Error)>,
+}
 
-    async fn authorize_request(&self, _request: &engine::Request) -> Result<ExecutionAuth, AuthError>;
+impl<Executor, Cache, Context, Response, Error> Gateway<Executor, Cache, Context, Response, Error>
+where
+    Error: From<Error> + std::error::Error + Send + 'static,
+    Context: RequestContext,
+    Response: self::Response<Error = Error>,
+    Cache: self::Cache<Value = engine::Response> + 'static,
+    Executor: self::Executor<Context = Context, Error = Error, Response = Response> + 'static,
+{
```

Comment:
`Response` & `Error` come from the `Executor` so I think you could ditch them as generic params:

```suggestion
impl<Executor, Cache, Contextr> Gateway<Executor, Cache, Context>
where
    Executor: self::Executor<Context = Context> + 'static,
    Executor::Response: self::Response<Error = Executor::Error>,
    Executor::Error: From<Error> + std::error::Error + Send + 'static,
    Context: RequestContext,
    Cache: self::Cache<Value = engine::Response> + 'static,
{
```

Reply:
good catch!